pg_bm25/src/manager/mod.rs

```diff
@@ -143,3 +143,215 @@ impl Manager {
+    }
+
+    #[test]
+    fn test_add_doc_address() {
+        let first_item_ptr = ItemPointerData {
+            ip_blkid: BlockIdData {
+                bi_hi: 10,
+                bi_lo: 0,
+            },
+            ip_posid: 8,
+        };
+        let second_item_ptr = ItemPointerData {
+            ip_blkid: BlockIdData {
+                bi_hi: 88,
+                bi_lo: 22,
+            },
+            ip_posid: 3,
+        };
+        let first_ctid = item_pointer_get_both(first_item_ptr);
+        let second_ctid = item_pointer_get_both(second_item_ptr);
+
+        let first_doc_address = DocAddress::new(0, 1);
+        let second_doc_address = DocAddress::new(0, 2);
+
+        let manager = get_fresh_executor_manager();
+        manager.add_doc_address(first_ctid, first_doc_address);
+        manager.add_doc_address(second_ctid, second_doc_address);
+
+        let mut expected: HashMap<BlockInfo, DocAddress> = HashMap::new();
+        expected.insert((655360, 8), first_doc_address);
+        expected.insert((5767190, 3), second_doc_address);
+
+        assert_eq!(&expected, manager.doc_addresses.as_mut().unwrap());
+        assert_eq!(manager.doc_addresses.as_mut().unwrap().len(), 2);
+    }
+
+    fn prepare_schema() -> tantivy::Result<(Schema, Searcher, Field)> {
+        let mut schema_builder = Schema::builder();
+        let title = schema_builder.add_text_field("title", TEXT);
+        let schema = schema_builder.build();
+        let index = Index::create_in_ram(schema.clone());
+
+        {
+            let mut index_writer = index.writer(3_000_000)?;
+            index_writer.add_document(doc!(
+                title => "The Name of the Wind",
+            ))?;
+            index_writer.add_document(doc!(
+                title => "The Diary of Muadib",
+            ))?;
+            index_writer.add_document(doc!(
+                title => "A Dairy Cow",
+            ))?;
+            index_writer.add_document(doc!(
+                title => "The Diary of a Young Girl",
+            ))?;
+            index_writer.commit()?;
+        }
+
+        let reader = index.reader()?;
+        let searcher = reader.searcher();
+        Ok((schema, searcher, title))
+    }
+
+    #[test]
+    fn test_add_snippet_generators() -> tantivy::Result<()> {
+        let (schema, searcher, title) = prepare_schema()?;
+        let query: Box<dyn Query> = Box::new(RegexQuery::from_pattern("d[ai]{2}ry", title)?);
+
+        let manager = get_fresh_executor_manager();
+        manager.add_snippet_generators(&searcher, &schema, &query, Some(3));
+        let snippet_generators = manager.snippet_generators.as_mut().unwrap();
+
+        assert_eq!(snippet_generators.len(), 1);
+        assert!(snippet_generators.get("title").is_some());
+        assert!(snippet_generators.get("id").is_none());
+
+        Ok(())
+    }
+
+    #[test]
+    #[should_panic(expected = "failed to retrieve snippet generator for field: {field_name}")]
```

Comment:
I think we could probably leave off the `expected` string here, our error handling is still unstable.

Reply:
Okay, thanks for that.